### ISSUE TITLE
Issues with socialhome <-> ganggo federation

### ISSUE THREAD

**Issue by Zauberstuhl** [jaywink/socialhome#434]
related to https://github.com/ganggo/federation/issues/12

GangGo extracts the public key from the Hcard.

Can you verify that for this account https://socialhome.network/hcard/users/206da4d1-0887-4b2b-bb91-3f1cbc0006df
the public key is correct and used for encryption?

Cause my server log tells me that GangGo has a problem with decrypting your messages:
```
ERROR 2018/02/19 11:34:55    app  logger.go:59: [[ [[crypto/rsa: decryption error]] ]] 
ERROR 2018/02/19 11:34:55    app receiver.go:115: Cannot parse encrypted request           ip=172.17.0.1 path=/receive/users/c6c14f1d3e20a35ffdb7b6ce004115ee method=POST action=Receiver.Private namespace=App\\ error="crypto/rsa: decryption error" 
```

---

<del>Also the public key from https://jasonrobinson.me/hcard/users/da323072-e566-4b6e-a8b6-b2bda0c6ace4 and https://jasonrobinson.me/webfinger?q=acct:jaywink@jasonrobinson.me differ!</del>

<del>It doesn't matter for ganggo but maybe for other services!</del>

**Comment by jaywinky** [comment]
> Also the public key from https://jasonrobinson.me/hcard/users/da323072-e566-4b6e-a8b6-b2bda0c6ace4 and https://jasonrobinson.me/webfinger?q=acct:jaywink@jasonrobinson.me differ!

You scared me for a moment :D But no, the legacy webfinger XML key is just encoded in base64, as per spec. It used to be [here](https://wiki.diasporafoundation.org/Federation#Diaspora_Public_Key) but it seems the diaspora team decided to remove the old page completely so I can't link to it.

Regarding the verification, yes that is the case. Though note that socialhome.network runs an older version without my latest changes. Though, these changes did not really touch the signing parts, only delivery and private JSON encryption (which is new). Since we're talking about public content this shouldn't matter, but you wont be able to test against socialhome.network since it doesn't have the JSON webfinger ;) I will deploy there as soon as I fix a few things in master re private delivery.

**Comment by jaywinky** [comment]
> ERROR 2018/02/19 11:34:55    app receiver.go:115: Cannot parse encrypted request           ip=172.17.0.1 path=/receive/users/c6c14f1d3e20a35ffdb7b6ce004115ee method=POST action=Receiver.Private namespace=App\\ error="crypto/rsa: decryption error" 

Ah, that seems to be a private message, probably a follow entity. That doesn't work yet, broken in master. But it seems the public deliveries are broken too somehow?

**Comment by Zauberstuhl** [comment]
@jaywink can you create a test account for jasonrobinson for me =? cause I think the registration is closed.

regarding public posts; thats the same issue I guess `expected element type <me:env> but have <diaspora>`:
```
DEBUG 2018/02/19 13:07:35    app receiver.go:49: received publicly                        ip=172.17.0.1 path=/receive/public method=POST action=Receiver.Public namespace=App\\ message="<diaspora xmlns:me=\"http://salmon-protocol.org/ns/magic-env\" xmlns=\"https://joindiaspora.com/protocol\">\n  <header>\n    <author_id>zauberstuhl@socialhome.network</author_id>\n  </header>\n  <me:env>\n    <me:encoding>base64url</me:encoding>\n    <me:alg>RSA-SHA256</me:alg>\n    <me:data type=\"application/xml\">PFhNTD4KICA8cG9zdD4KICAgIDxzdGF0dXNfbWVzc2FnZT4KICAgICAgPHJhd19tZXNzYWdlPnBpbmcgMTIzNDwvcmF3X21lc3NhZ2U-CiAgICAgIDxndWlkPjcyNDljNDliLTRmNzUtNDdjZS1hYzVlLTNkNmNlNDA0NTllMDwvZ3VpZD4KICAgICAgPGRpYXNwb3JhX2hhbmRsZT56YXViZXJzdHVobEBzb2NpYWxob21lLm5ldHdvcms8L2RpYXNwb3JhX2hhbmRsZT4KICAgICAgPHB1YmxpYz50cnVlPC9wdWJsaWM-CiAgICAgIDxjcmVhdGVkX2F0PjIwMTgtMDItMTlUMTM6MDc6MTZaPC9jcmVhdGVkX2F0PgogICAgICA8cHJvdmlkZXJfZGlzcGxheV9uYW1lPlNvY2lhbGhvbWU8L3Byb3ZpZGVyX2Rpc3BsYXlfbmFtZT4KICAgIDwvc3RhdHVzX21lc3NhZ2U-CiAgPC9wb3N0Pgo8L1hNTD4K</me:data>\n    <me:sig>F0CBuT-w5x7WFDp8PSCUIOXbXmCw7gPUVyh9yoEW7KCjFEPKm1I1JIdorG76aDASr7EBqw1SY_w7MmOy40QkBzcEVgF18lJaA1SGS1DihwYx2wfFONVE3AdbN8BMKelW50mzwb6MZX0eYA7Lu-hII8LW9zxp9NebfnpWvRWi9zLUYNxyetrK6aZyQYl5ajG5EuVayUfmF-f6rKENYiQWiUbF5qAi_Gstks8Ni9PkESmEZcAQR9pyUuamyfHd0vyFjL_NdtOdzwM6WUXJ1_5V2nJrJKcpwcgqARJ3vJCExKzxMuM5Tug6UQzbbS3_7ZJanI_qy1SlZIlu2Z2MPYK-qBdpf-tteHERicXg-MkrifLAi5ZB7821nAQ6FWwXgZZ_xGvf8EoozxEd7yznd592J2xIZ-c4_AS2Ydtqb2mIxdXlS1HM9aUHyB8tjy5f2biMa99uMOEcMtk0pwpcwo2kavyaxoPQUTS5GgNxECgx7Kl8bXEIjfz_qXea9RZNRCf3qj4_Tql_LAoL1ICiHEur-3iSo84wrfrh3sGs2RmcVAjngba3A8uRBM8s437skGLwRmNLPwFWDVrpfjMn8sHiUFYFu2S4GCH1UtjBciynm7nz4ok-x6KMypCYUMaXxRNPlsJ4e1nWSp2Imk6jwcLwGDZDoabbZG7D16IhTwTtW-U=</me:sig>\n  </me:env>\n</diaspora>\n" 
ERROR 2018/02/19 13:07:35    app  logger.go:59: [[ [[expected element type <me:env> but have <diaspora>]] ]] 
```

**Comment by jaywink** [comment]
> regarding public posts; thats the same issue I guess expected element type <me:env> but have <diaspora>:

socialhome.network uses the old delivery. jasonrobinson.me the new without `<diaspora>`

**Comment by Zauberstuhl** [comment]
ok I think I found it.
```
<status_message>
  <raw_message>hello world</raw_message>
  <guid>651d35c5-06aa-48a5-9d29-3cf9a31fea1b</guid>
  <diaspora_handle>zauberstuhl@jasonrobinson.me</diaspora_handle>
  <public>true</public><created_at>2018-02-19T13:38:54Z</created_at>
  <provider_display_name>Socialhome</provider_display_name>
</status_message>
```

which is still the legacy layout. `diaspora_handle` becomes `author` (see https://diaspora.github.io/diaspora_federation/entities/status_message.html)

**Comment by jaywink** [comment]
Bummer, I was supposed to switch fully to sending only with the latest attributes at the same time but forgot - and things work towards diaspora and others due to the legacy ones they support still. I will switch asap hopefully today evening! Thanks for testing!

**Comment by jaywink** [comment]
Federation ref: https://github.com/jaywink/federation/issues/59

**Comment by jaywink** [comment]
Closed by #440 \o/
